### ISSUE TITLE
Using seed breaks .integer

### ISSUE THREAD

**Issue by zachdsawyer** [chancejs/chancejs#422]
I'm using 1.0.18.
Reproduce with this: `new Chance(Math.random()).integer({min:1,max:3})`
This returns a value of `2` 100% of the time.

**Comment by somejeff** [comment]
The correct syntax for the Chance constructor is to pass in an integer (if you want a consistent seed).

Passing in ```Math.random()``` is incorrect because it's a decimal number between 0 and 1, therefore, by chance, 2 is always returned.

In your example above, you shouldn't pass anything in.
``` javascript
> new Chance().integer({min:1,max:3})
3
> new Chance().integer({min:1,max:3})
1
> new Chance().integer({min:1,max:3})
3
> new Chance().integer({min:1,max:3})
2
> new Chance().integer({min:1,max:3})
2
> new Chance().integer({min:1,max:3})
3
> new Chance().integer({min:1,max:3})
2
```

Under the hood, this is equivalent to:
```javascript
new Chance(Math.floor(Math.random()*Math.pow(10,13))).integer({min:1,max:3})
```
This is like saying "Pick a whole number between 0 and 10000000000000"

**Comment by zachdsawyer** [comment]
Given that you can even pass in a STRING as a seed,  and the docs say nothing about requiring an integer, a correct assumption is that any number will work.  Therefore I'm still considering this a bug.

**Comment by gerardolima** [comment]
hey, @somejeff, although the behaviour is consistent with spec, @zachdsawyer has a point about how it's easy to use a decimal number, instead of an integer. Maybe the constructor should check for numeric seeds and throw error in case of not integer numbers? Otherwise, could that be avoided at all by always requiring/using seed as a string?

**Comment by somejeff** [comment]
Hey @gerardolima, here's my feedback as to why `Math.random()` as a seed isn't optimal, but also technically supported.

TL;DR: The built-in MersenneTwister takes the seed, unsigned right shifted, and goes from there.
Anything between 0 and 1 becomes 0.

----

The built-in MersenneTwister uses the seed to populate an array of pseudorandom numbers.
It's expecting a seed between 0 and 10_000_000_000_000

Let's pick a number from 1 to 10, using seeds 0 to 4.

```javascript
new Chance(0).integer({min:1,max:10})
```
| Seed  | First few elements of the array | First call to `mt.random()` to choose an int | Result | 
| --- | --- | --- | --- |
| 0 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 1 |  `[1, 1812433254, ...]` | 0.4170219984371215 | 5 |
| 2 |  `[2, 3624866507, ...]` | 0.43599490262567997 | 5 |
| 3 |  `[3, 1142332464, ...]` | 0.5507979043759406 | 6 |
| 4 |  `[4, 2954765717, ...]` | 0.9670298383571208 | 10 |

Initializing the MT with such close integers and only providing a field of 1-10 for `integer()` to choose (**once**), returns 6,5,5,10.

---- 

Narrow the field to `{min:1,max:3}` is like shrinking the dart board down to only 3 options.

 ```javascript
new Chance(0).integer({min:1,max:3})
```
| Seed  | First few elements of the array | First call to `mt.random()` to choose an int | Result | 
| --- | --- | --- | --- |
| 0 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 2 |
| 1 |  `[1, 1812433254, ...]` | 0.4170219984371215 | 2 |
| 2 |  `[2, 3624866507, ...]` | 0.43599490262567997 | 2 |
| 3 |  `[3, 1142332464, ...]` | 0.5507979043759406 | 2 |
| 4 |  `[4, 2954765717, ...]` | 0.9670298383571208 | 3 |

----

Let's see what happens if you pass a seed between 0 and 1, essentially populating the array with zero as the first entry and always returning the same pseudorandom value.


 ```javascript
new Chance(0.x).integer({min:1,max:10})
```
| Seed  | First few elements of the array | First call to `mt.random()` to choose an int | Result | 
| --- | --- | --- | --- |
| 0 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.1 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.2 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.3 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.4 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.5 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.6 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.7 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.8 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.9 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 0.99999999 |  `[0, 1, 1812433255, ...]` | 0.548813502304256 | 6 |
| 1 |  `[1, 1812433254, ...]` | 0.4170219984371215 | 5 |

The same would apply to `{min:1,max:3}` and always return 2.

---
Could the MersenneTwister be improved to support real numbers between 0 and 1 instead of ints?  
Sure, but any change to the twister is a breaking change for all those expecting patterns from their fixed seeds.

Could the constructor validate and reject real numbers?
Also a breaking change for those passing a real number with a larger field, such as:
`new Chance(Math.random()*10_000_000_000_000);`

String seeds?
If a String is passed, it's converted into an integer using each letter of the string.  😐

--- 

Recommendation?  

Take advantage of the full 32 bits for a seed and stay away from 0 -> 1 and the other limit too: 
```javascript
new Chance(Math.random() * Infinity);  // https://xkcd.com/221/
```

**Comment by victorquinn** [comment]
Fantastic analysis here @somejeff , thank you!

Yes unfortunately for those taking issue with this, this is expected behavior, we should update the docs to describe it as such and clarify.

**Comment by gerardolima** [comment]
hey again, @somejeff, I understand your explanation and I agree that keeping a stable interface is precious for any library.

My comment was about improving the ergonomics of the API by avoiding what looks to be a rather silly, but also one easy mistake to make. As a suggestion, Chance could trigger an error (or log a warning message, maybe?), when `seed % 1 !== 0`.

That said, I understand that this might break the behaviour in some cases -- and maybe unfold some unintended usage of `Chance(Math.random())` -- so it's a balance and it's obviously your call. Thank you for your good work!
